iamroot17A: 16주차 스터디

JSYoo5B opened the pull request:
* 2020년 12월 12일 (15:00~22:00)    
* 코드 버전: mainline 5.9           
* init/main.c start_kernel() 분석 중
  * local_irq_disable() 분석        
  * boot_cpu_init() 분석            
  * page_address_init() 분석        
  * pr_notice() 분석 (linux_banner) 
* ARM64에서 구현된 atomic operation 관련 내용 분석 
  LL/SC, LSE방식으로 구현된 atomic operation 확인  
  bitops에서 구현된 CAS(Compare-And-Swap) 방식 확인

JSYoo5B committed:
iamroot17A: 16주차 오후 스터디

* 2020년 12월 12일 (15:00~18:00)
* 코드 버전: mainline 5.9
* init/main.c start_kernel() 분석 중
  * local_irq_disable() 분석
  * boot_cpu_init() 분석
  * page_address_init() 분석
  * pr_notice() 분석 (linux_banner)

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B committed:
iamroot17A: 16주차 저녁 스터디

* 2020년 12월 12일 (19:00~22:00)
* 코드 버전: mainline 5.9
* ARM64에서 구현된 atomic operation 관련 내용 분석
  LL/SC, LSE방식으로 구현된 atomic operation 확인
  bitops에서 구현된 CAS(Compare-And-Swap) 방식 확인

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B commented:
잘못된 부분이 있나 확인 부탁드립니다.

별 문제 없으면 며칠 뒤 알아서 merge 해 놓도록 하겠습니다.

dngthe93 commented on arch/arm64/include/asm/smp.h:
```diff
@@ -40,6 +44,11 @@ DECLARE_PER_CPU_READ_MOSTLY(int, cpu_number);
  */
+/*; Iamroot17A 2020.Dec.12 #2.2.3
+ *;
+ *; raw_cpu_ptr의 macro에서 나온 주소 값을 pointer indirect 연산을 통해
```
> pointer indirect 연산을 통해

라는 문장 보다는,
> pointer dereference operator를 통해

라는 표현을 쓰는게 어떨까요?

JSYoo5B replied:
pointer dereference operator라는 명칭이 있었군요.

생각나는 단어가 indirect 연산이었습니다.

dngthe93 commented on include/linux/cpumask.h:
```diff
@@ -14,6 +14,11 @@
+/*; Iamroot17A 2020.Dec.12 #2.3.2
+ *;
+ *; CPU의 상태를 표현하기 위해 BITMAP으로 표현한다.
+ *; (NR_CPUS의 갯수만큼 bit를 표현하기 위한 unsigned long 배열)
```
`갯수` 보다는 `개수`가 더 적절할 것 같습니다

JSYoo5B replied:
개수가 표준어였군요. 아래 리뷰 부분도 같이 수정해 놓도록 하겠습니다.

dngthe93 commented on include/linux/types.h:
```diff
@@ -7,6 +7,11 @@
 
+/*; Iamroot17A 2020.Dec.12 #2.3.3
+ *;
+ *; BITS_TO_LONGS는 해당 bit를 표현하기 위해 필요한 unsigned long의 갯수를
```
`갯수` 보다는 `개수`가 더 적절할 것 같습니다

dngthe93 commented on kernel/cpu.c:
```diff
@@ -2473,6 +2473,14 @@ void set_cpu_online(unsigned int cpu, bool online)
+		 *;
+		 *; set_cpu_active(), set_cpu_present(), set_cpu_possible()은
+		 *; cpumast_set_cpu로 현재 state에 관계없이 overwrite를 하지만
+		 *; set_cpu_online()은 __num_online_cpus의 갯수 조절을 위해
```
`갯수` 보다는 `개수`가 더 적절할 것 같습니다

JSYoo5B replied:
여기는 개수보단 값 이란 표현이 더 명학해보이네요.

dngthe93 commented on arch/arm64/include/asm/lse.h:
```diff
@@ -25,6 +25,18 @@ static inline bool system_uses_lse_atomics(void)
 
+/*; Iamroot17A 2020.Dec.12 #5.1
+ *;
+ *; 현재 빌드 config 설정에 따라 LSE방식 혹은 LL/SC방식을 선택하게 된다.
```
> LSE방식 혹은 LL/SC방식을 선택하게 된다.

보다는

> (LSE or LL/SC) 방식 혹은 (LL/SC) 방식을 선택하게 된다.

가 좀 더 명확하게 이해가 잘 될 것 같습니다.

JSYoo5B replied:
정리하다가 내용 생략이 좀 많이 되었네요.
말씀하신게 좀 더 명확할 것 같습니다.

dngthe93 reviewed (dismissed):
정리 감사합니다.  
나름대로 수정할 사항 리뷰해보았습니다.

(리뷰에 있는 Request changes 기능이 궁금하여 체크해보았습니다.)

dngthe93 commented on mm/highmem.c:
```diff
@@ -475,6 +475,12 @@ void __init page_address_init(void)
+	 *;
+	 *; struct page_address_slot을 사용한 배열 struct_address_htable의
+	 *; list와 spinlock을 초기화한다.
+	 *; 그런데 왜 여기에서는 hlist가 아니라 list를 사용하는지 의문
```
`page_address_htable` 자료구조는 tail에다가 추가하는 기능이 필요했나봅니다.  
정말 그러한지는 나중에 분석해보면 좋을 것 같은데... 이후 진도에서 나올지 궁금하네요.

JSYoo5B replied:
저희가 분석할때도 흐름상 원래 아무 일도 안 하는 부분이었기때문에 과연 나타날지 의문이네요.

해당 부분에 대하여 blame으로 확인해 보니 제일 첫 commit이 나오는군요.
더 이전 기록이라서 정확한 원인은 알 수 없네요.

혹시나 hlist가 이후 개발되어 아직 수정되지 않은 것인가 의심해봤는데
hlist도 최초 commit에서 나타나서 알 수가 없네요.

tail에서 추가하는 기능이 없다면 refactoring으로 기여를 할 수도 있어 보입니다.

JSYoo5B committed:
typo: 맞춤법 변경

* 갯수 -> 개수로 변경
  <한글 맞춤법> 사이시옷 규정에 따라 ‘자리’와 ‘수(數)’가 결합하여
  만들어진 말은 ‘자릿수’와 같이 적습니다. 하지만 ‘곳간(庫間),
  셋방(貰房), 숫자(數字), 찻간(車間), 툇간(退間), 횟수(回數)’를 제외한
  한자어에는 사이시옷을 붙이지 않습니다.
* Documentation/translations/ko_KR/memory-barriers.txt에서도
  발견되었으나 현재 분석에 영향을 주는 부분이 아니므로 생략함

Signed-off-by: JaeSang Yoo <jsyoo5b@gmail.com>

JSYoo5B commented:
기존 commit에 존재하던 맞춤법 문제도 같이 수정합니다.
